phd-tests/framework/src/disk/file.rs

```diff
@@ -4,22 +4,84 @@
 
 //! Abstractions for disks with a raw file backend.
 
-use std::path::{Path, PathBuf};
-
+use camino::{Utf8Path, Utf8PathBuf};
 use propolis_client::types::{FileStorageBackend, StorageBackendV0};
-use tracing::{error, info};
+use tracing::{debug, error, warn};
 use uuid::Uuid;
 
-use crate::guest_os::GuestOsKind;
+use crate::{guest_os::GuestOsKind, zfs::ClonedFile as ZfsClonedFile};
+
+/// Describes the method used to create the backing file for a file-backed disk.
+#[derive(Debug)]
+enum BackingFile {
+    /// The disk is a ZFS clone of the original artifact.
+    Zfs(ZfsClonedFile),
+
+    /// The disk is a hard copy of the original artifact.
+    HardCopy(Utf8PathBuf),
+}
+
+impl BackingFile {
+    /// Creates a new backing file from the artifact with the supplied
+    /// `artifact_path`. If possible, this routine will create a ZFS clone of
+    /// the dataset containing the file; otherwise it will fall back to creating
+    /// a hard copy of the original artifact.
+    fn create_from_source(
+        artifact_path: &Utf8Path,
+        data_dir: &Utf8Path,
+    ) -> anyhow::Result<Self> {
+        if let Ok(file) = ZfsClonedFile::create_from_path(artifact_path) {
+            return Ok(Self::Zfs(file));
+        }
+
+        warn!(%artifact_path,
+              "failed to make ZFS clone of backing artifact, will copy it");
```

Comment:
nit, not a big deal: it might be nice to include the error in the warning?
```suggestion
        match ZfsClonedFile::create_from_path(artifact_path) {
            Ok(file) => return Ok(Self::Zfs(file)),
            Err(error) => warn!(
                %artifact_path,
                %error,
               "failed to make ZFS clone of backing artifact, will copy it",
            ),
        }

```

Reply:
Done in 409c7bd.